src/audio_pixel.cpp

```diff
@@ -0,0 +1,276 @@
+#include "audio_pixel.h"
+#define project nullptr
+
+// audio pixel functions 
+
+double linear_interp(double x, double x1, double x2, double y1, double y2) {
+    // edge cases 
+    if ((x2-x1) == 0)
+        return x1;
+    // interpolation in time domain
+    return (((x2 - x) / (x2 - x1)) * y1) + ((x - x1) / (x2 - x1)) * y2;
+}
+
+double closest_val(double val1, double val2, double target) {
+    return abs(target - val1) <= abs(target - val2) ? val1 : val2;
+}
+
+template<typename T>
+const std::vector<T> get_view(std::vector<T> const &parent, int start, int end)
+{
+    // todo: make sure start and  end indices are valid
+    auto startptr = v.cbegin() + start;
+    auto endptr = v.cbegin() + end;
+ 
+    std::vector<T> vec(startptr, endptr);
+    return vec;
+}
+
+// ************* audio_pixel_block_t ****************
+
+bool audio_pixel_block_t::flush() const {
+    json j;
+    to_json(j);
+    std::string resource_path = GetResourcePath();
+    std::ofstream ofs;
+    ofs.open(resource_path + "/kiwi-block.json");
+
+    if (!ofs.is_open()){
+        return false;
+    }
+    
+    ofs << std::setw(4) << j << std::endl;
+    return true;
+}
+
+void audio_pixel_block_t::to_json(json& j) const {
+    j[std::to_string(m_pix_per_s)] = m_channel_pixels;
+}
+
+void audio_pixel_block_t::update(std::vector<double>& sample_buffer, int num_channels, int sample_rate) {
+    // constructs an audio pixel block using interleaved sample buffer and num channels
+    // clear the current set of audio pixels 
```

Comment:
move me above the function decl in the header!

Reply:
i got u i will also add this to the style guide